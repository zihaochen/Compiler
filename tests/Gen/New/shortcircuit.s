.data
	_printf_cnt: .space 4
	str_1: .space 48
	str_2: .space 124
.text
main:
	move $fp, $sp     # start using memory here
	move $gp, $sp     # set global pointer (unused)
	subu $sp, $sp, 4
	li $t0, 0
	sw $t0, _printf_cnt
	li $t0, 101
	la $t1, str_1
	sw $t0, 0($t1)
	li $t0, 120
	la $t1, str_1
	sw $t0, 4($t1)
	li $t0, 112
	la $t1, str_1
	sw $t0, 8($t1)
	li $t0, 108
	la $t1, str_1
	sw $t0, 12($t1)
	li $t0, 111
	la $t1, str_1
	sw $t0, 16($t1)
	li $t0, 100
	la $t1, str_1
	sw $t0, 20($t1)
	li $t0, 101
	la $t1, str_1
	sw $t0, 24($t1)
	li $t0, 32
	la $t1, str_1
	sw $t0, 28($t1)
	li $t0, 37
	la $t1, str_1
	sw $t0, 32($t1)
	li $t0, 100
	la $t1, str_1
	sw $t0, 36($t1)
	li $t0, 10
	la $t1, str_1
	sw $t0, 40($t1)
	li $t0, 0
	la $t1, str_1
	sw $t0, 44($t1)
	li $t0, 87
	la $t1, str_2
	sw $t0, 0($t1)
	li $t0, 111
	la $t1, str_2
	sw $t0, 4($t1)
	li $t0, 119
	la $t1, str_2
	sw $t0, 8($t1)
	li $t0, 44
	la $t1, str_2
	sw $t0, 12($t1)
	li $t0, 32
	la $t1, str_2
	sw $t0, 16($t1)
	li $t0, 99
	la $t1, str_2
	sw $t0, 20($t1)
	li $t0, 111
	la $t1, str_2
	sw $t0, 24($t1)
	li $t0, 110
	la $t1, str_2
	sw $t0, 28($t1)
	li $t0, 103
	la $t1, str_2
	sw $t0, 32($t1)
	li $t0, 114
	la $t1, str_2
	sw $t0, 36($t1)
	li $t0, 97
	la $t1, str_2
	sw $t0, 40($t1)
	li $t0, 116
	la $t1, str_2
	sw $t0, 44($t1)
	li $t0, 117
	la $t1, str_2
	sw $t0, 48($t1)
	li $t0, 108
	la $t1, str_2
	sw $t0, 52($t1)
	li $t0, 97
	la $t1, str_2
	sw $t0, 56($t1)
	li $t0, 116
	la $t1, str_2
	sw $t0, 60($t1)
	li $t0, 105
	la $t1, str_2
	sw $t0, 64($t1)
	li $t0, 111
	la $t1, str_2
	sw $t0, 68($t1)
	li $t0, 110
	la $t1, str_2
	sw $t0, 72($t1)
	li $t0, 115
	la $t1, str_2
	sw $t0, 76($t1)
	li $t0, 33
	la $t1, str_2
	sw $t0, 80($t1)
	li $t0, 10
	la $t1, str_2
	sw $t0, 84($t1)
	li $t0, 116
	la $t1, str_2
	sw $t0, 88($t1)
	li $t0, 49
	la $t1, str_2
	sw $t0, 92($t1)
	li $t0, 32
	la $t1, str_2
	sw $t0, 96($t1)
	li $t0, 61
	la $t1, str_2
	sw $t0, 100($t1)
	li $t0, 32
	la $t1, str_2
	sw $t0, 104($t1)
	li $t0, 37
	la $t1, str_2
	sw $t0, 108($t1)
	li $t0, 100
	la $t1, str_2
	sw $t0, 112($t1)
	li $t0, 10
	la $t1, str_2
	sw $t0, 116($t1)
	li $t0, 0
	la $t1, str_2
	sw $t0, 120($t1)
	j _main
_isdecdigit:
	sw $ra, ($sp)
	li $t2, 48
	lw $t1, 4($sp)
	sge $t0, $t1, $t2
	sw $t0, -4($sp)
	lw $t0, -4($sp)
	bne $t0, $0, L3
	b L2
L3:
	li $t2, 57
	lw $t1, 4($sp)
	sle $t0, $t1, $t2
	sw $t0, -8($sp)
	lw $t0, -8($sp)
	bne $t0, $0, L1
	b L2
L1:
	li $v0, 1
	sw $v0, 0($sp)
	jr $ra
L2:
	li $v0, 0
	sw $v0, 0($sp)
	jr $ra
	jr $ra
_getint:
	sw $ra, ($sp)
L5:
	subu $sp, $sp, 60
	jal _getchar
	addi $sp, $sp, 60
	sw $v0, -16($sp)
	lw $ra, ($sp)
	lw $t0, -16($sp)
	sw $t0, -4($sp)
	lw $t0, -4($sp)
	sw $t0, -12($sp)
	lw $t0, -12($sp)
	sw $t0, -60($sp)
	subu $sp, $sp, 64
	jal _isdecdigit
	addi $sp, $sp, 64
	sw $v0, -20($sp)
	lw $ra, ($sp)
	lw $t0, -20($sp)
	bne $t0, $0, L4
	b L6
L6:
	b L5
	b L5
L4:
	li $t2, 48
	lw $t1, -4($sp)
	sub $t0, $t1, $t2
	sw $t0, -28($sp)
	lw $t0, -28($sp)
	sw $t0, -8($sp)
	lw $t0, -8($sp)
	sw $t0, -24($sp)
L8:
	subu $sp, $sp, 60
	jal _getchar
	addi $sp, $sp, 60
	sw $v0, -36($sp)
	lw $ra, ($sp)
	lw $t0, -36($sp)
	sw $t0, -4($sp)
	lw $t0, -4($sp)
	sw $t0, -32($sp)
	lw $t0, -32($sp)
	sw $t0, -60($sp)
	subu $sp, $sp, 64
	jal _isdecdigit
	addi $sp, $sp, 64
	sw $v0, -40($sp)
	lw $ra, ($sp)
	lw $t0, -40($sp)
	bne $t0, $0, L9
	b L7
L9:
	li $t2, 10
	lw $t1, -8($sp)
	mul $t0, $t1, $t2
	sw $t0, -48($sp)
	lw $t2, -4($sp)
	lw $t1, -48($sp)
	add $t0, $t1, $t2
	sw $t0, -52($sp)
	li $t2, 48
	lw $t1, -52($sp)
	sub $t0, $t1, $t2
	sw $t0, -56($sp)
	lw $t0, -56($sp)
	sw $t0, -8($sp)
	lw $t0, -8($sp)
	sw $t0, -44($sp)
	b L8
L7:
	lw $v0, -8($sp)
	sw $v0, -8($sp)
	jr $ra
	jr $ra
_explode:
	sw $ra, ($sp)
	la $t0, str_1
	sw $t0, -16($sp)
	lw $t0, 4($sp)
	bne $t0, $0, L11
	b L10
L11:
	li $t0, 0
	sw $t0, -4($sp)
	lw $t0, -4($sp)
	sw $t0, -8($sp)
L13:
	li $t2, 1000
	lw $t1, -4($sp)
	slt $t0, $t1, $t2
	sw $t0, -12($sp)
	lw $t0, -12($sp)
	bne $t0, $0, L14
	b L12
L14:
	lw $t0, -16($sp)
	sw $t0, -32($sp)
	lw $t0, -4($sp)
	sw $t0, -36($sp)
	li $t0, 2
	sw $t0, _printf_cnt
	subu $sp, $sp, 40
	jal _printf
	addi $sp, $sp, 40
	sw $v0, -20($sp)
	lw $ra, ($sp)
	li $t2, 1
	lw $t1, -4($sp)
	add $t0, $t1, $t2
	sw $t0, -28($sp)
	lw $t0, -28($sp)
	sw $t0, -4($sp)
	lw $t0, -4($sp)
	sw $t0, -24($sp)
	b L13
L12:
L10:
	li $v0, 1
	sw $v0, 0($sp)
	jr $ra
	jr $ra
_main:
	sw $ra, ($sp)
	la $t0, str_2
	sw $t0, -100($sp)
	subu $sp, $sp, 124
	jal _getint
	addi $sp, $sp, 124
	sw $v0, -8($sp)
	lw $ra, ($sp)
	li $t2, 1
	lw $t1, -8($sp)
	add $t0, $t1, $t2
	sw $t0, -12($sp)
	lw $t0, -12($sp)
	sw $t0, -4($sp)
	subu $sp, $sp, 124
	jal _getint
	addi $sp, $sp, 124
	sw $v0, -20($sp)
	lw $ra, ($sp)
	li $t2, 2
	lw $t1, -20($sp)
	add $t0, $t1, $t2
	sw $t0, -24($sp)
	lw $t0, -24($sp)
	sw $t0, -16($sp)
	subu $sp, $sp, 124
	jal _getint
	addi $sp, $sp, 124
	sw $v0, -32($sp)
	lw $ra, ($sp)
	li $t2, 3
	lw $t1, -32($sp)
	add $t0, $t1, $t2
	sw $t0, -36($sp)
	lw $t0, -36($sp)
	sw $t0, -28($sp)
	li $t0, 100
	sw $t0, -40($sp)
L16:
	li $t2, 0
	lw $t1, -40($sp)
	sgt $t0, $t1, $t2
	sw $t0, -44($sp)
	lw $t0, -44($sp)
	bne $t0, $0, L17
	b L15
L17:
	li $t2, 0
	lw $t1, -4($sp)
	sgt $t0, $t1, $t2
	sw $t0, -48($sp)
	lw $t0, -48($sp)
	bne $t0, $0, L19
	b L21
L21:
	lw $t0, -4($sp)
	sw $t0, -124($sp)
	subu $sp, $sp, 128
	jal _explode
	addi $sp, $sp, 128
	sw $v0, -52($sp)
	lw $ra, ($sp)
	lw $t0, -52($sp)
	bne $t0, $0, L19
	b L20
L19:
	li $t2, 1
	lw $t1, -16($sp)
	sgt $t0, $t1, $t2
	sw $t0, -56($sp)
	lw $t0, -56($sp)
	bne $t0, $0, L26
	b L25
L26:
	li $t2, 2
	lw $t1, -28($sp)
	sgt $t0, $t1, $t2
	sw $t0, -60($sp)
	lw $t0, -60($sp)
	bne $t0, $0, L23
	b L25
L25:
	lw $t0, -16($sp)
	sw $t0, -124($sp)
	subu $sp, $sp, 128
	jal _explode
	addi $sp, $sp, 128
	sw $v0, -64($sp)
	lw $ra, ($sp)
	lw $t0, -64($sp)
	bne $t0, $0, L23
	b L24
L23:
	li $t2, 0
	lw $t1, -16($sp)
	sle $t0, $t1, $t2
	sw $t0, -68($sp)
	lw $t0, -68($sp)
	bne $t0, $0, L30
	b L29
L30:
	lw $t0, -28($sp)
	sw $t0, -124($sp)
	subu $sp, $sp, 128
	jal _explode
	addi $sp, $sp, 128
	sw $v0, -72($sp)
	lw $ra, ($sp)
	lw $t0, -72($sp)
	bne $t0, $0, L28
	b L29
L28:
	lw $t0, -28($sp)
	sw $t0, -124($sp)
	subu $sp, $sp, 128
	jal _explode
	addi $sp, $sp, 128
	sw $v0, -76($sp)
	lw $ra, ($sp)
	b L27
L29:
	lw $t0, -16($sp)
	bne $t0, $0, L38
	b L37
L38:
	lw $t0, -28($sp)
	bne $t0, $0, L35
	b L37
L37:
	li $t0, 1000
	sw $t0, -124($sp)
	subu $sp, $sp, 128
	jal _explode
	addi $sp, $sp, 128
	sw $v0, -80($sp)
	lw $ra, ($sp)
	lw $t0, -80($sp)
	bne $t0, $0, L35
	b L36
L36:
	lw $t0, -4($sp)
	bne $t0, $0, L35
	b L33
L35:
	li $t0, 0
	bne $t0, $0, L34
	b L33
L34:
	li $t0, 1000
	sw $t0, -124($sp)
	subu $sp, $sp, 128
	jal _explode
	addi $sp, $sp, 128
	sw $v0, -84($sp)
	lw $ra, ($sp)
	lw $t0, -84($sp)
	bne $t0, $0, L32
	b L33
L32:
	li $t0, 1000
	sw $t0, -124($sp)
	subu $sp, $sp, 128
	jal _explode
	addi $sp, $sp, 128
	sw $v0, -88($sp)
	lw $ra, ($sp)
	b L31
L33:
	li $t0, 0
	sw $t0, -124($sp)
	subu $sp, $sp, 128
	jal _explode
	addi $sp, $sp, 128
	sw $v0, -92($sp)
	lw $ra, ($sp)
	lw $t0, -92($sp)
	bne $t0, $0, L41
	b L39
L41:
	li $t2, 1
	lw $t1, -40($sp)
	seq $t0, $t1, $t2
	sw $t0, -96($sp)
	lw $t0, -96($sp)
	bne $t0, $0, L40
	b L39
L40:
	lw $t0, -100($sp)
	sw $t0, -124($sp)
	lw $t0, -4($sp)
	sw $t0, -128($sp)
	li $t0, 2
	sw $t0, _printf_cnt
	subu $sp, $sp, 132
	jal _printf
	addi $sp, $sp, 132
	sw $v0, -104($sp)
	lw $ra, ($sp)
L39:
L31:
L27:
	b L22
L24:
	lw $t0, -16($sp)
	sw $t0, -124($sp)
	subu $sp, $sp, 128
	jal _explode
	addi $sp, $sp, 128
	sw $v0, -108($sp)
	lw $ra, ($sp)
L22:
	b L18
L20:
	lw $t0, -4($sp)
	sw $t0, -124($sp)
	subu $sp, $sp, 128
	jal _explode
	addi $sp, $sp, 128
	sw $v0, -112($sp)
	lw $ra, ($sp)
L18:
	li $t2, 1
	lw $t1, -40($sp)
	sub $t0, $t1, $t2
	sw $t0, -120($sp)
	lw $t0, -120($sp)
	sw $t0, -40($sp)
	lw $t0, -40($sp)
	sw $t0, -116($sp)
	b L16
L15:
	li $v0, 0
	sw $v0, 0($sp)
	jr $ra
	jr $ra

_printf:
	lw $t0, _printf_cnt
	mul $t0, $t0, 4
	add $t0, $t0, $sp
	lw $a1, ($t0)
	move $a2, $t0
	
_printf_loop:
	lb $a0, 0($a1)
	beq $a0, 0, _printf_end
	addu $a1, $a1, 4
	beq $a0, '%', _printf_fmt
	li $v0, 11
	syscall
	b _printf_loop

_printf_fmt:
	lb $a0, 0($a1)
	addu $a1, $a1, 4
	beq $a0, 'd', _printf_int
	beq $a0, 's', _printf_str
	beq $a0, 'c', _printf_char
	beq $a0, '0', _printf_width
	beq $a0, '.', _printf_width

_printf_int:
	subu $a2, $a2, 4
	lw $a0, 0($a2)
	li $v0, 1
	syscall
	b _printf_loop

_printf_str:
	subu $a2, $a2, 4
	lw $a3, 0($a2)
	b _printf_str_loop

_printf_str_loop:
	addu $a3, $a3, 4
	lw $a0, ($a3)
	beq $a0, 0, _printf_loop
	li $v0, 11
	syscall
	b _printf_str_loop

_printf_char:
	subu $a2, $a2, 4
	lw $a0, 0($a2)
	li $v0, 11
	syscall
	b _printf_loop

_printf_width:
	lb $t1, 0($a1)
	addu $a1, $a1, 4
	sub $t1, $t1, '1'
	addu $a1, $a1, 4
	subu $a2, $a2, 4
	lw $t0, 0($a2)
	blt $t0, 10, _printf_width_tmp
	subu $t1, $t1, 1
	blt $t0, 100, _printf_width_tmp
	subu $t1, $t1, 1
	blt $t0, 1000, _printf_width_tmp
	subu $t1, $t1, 1
	blt $t0, 10000, _printf_width_tmp
	subu $t1, $t1, 1
	blt $t0, 100000, _printf_width_tmp
	subu $t1, $t1, 1
	blt $t0, 1000000, _printf_width_tmp
	subu $t1, $t1, 1
	blt $t0, 10000000, _printf_width_tmp
	subu $t1, $t1, 1
	blt $t0, 100000000, _printf_width_tmp
	subu $t1, $t1, 1

_printf_width_tmp:
	li $a0, '0'
	beq $t1, 0, _printf_width_tmp2
	subu $t1, $t1, 1
	li $v0, 11
	syscall
	b _printf_width_tmp

_printf_width_tmp2:
	move $a0, $t0
	li $v0, 1
	syscall
	b _printf_loop

_printf_end:
	jr $ra

_getchar:
	li $v0, 12
	syscall
	jr $ra

_malloc_struct:
	lw $a0, 4($sp)
	li $v0, 9
	syscall
	move $t0, $v0
	li $v0, 9
	li $a0, 4
	syscall
	sw $t0, ($v0)
	jr $ra

_malloc:
	lw $a0, 4($sp)
	li $v0, 9
	syscall
	jr $ra
