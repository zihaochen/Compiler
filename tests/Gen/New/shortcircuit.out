_start: 
_printf_cnt = 0

_malloc: 

_getchar: 

_printf: 

_isdecdigit: 
Temp1 = ch GE 48
If Not Equal to Zero Temp1 Goto Label :3	
Goto label:2
Label :3	Temp2 = ch LE 57
If Not Equal to Zero Temp2 Goto Label :1	
Goto label:2
Label :1	Return BasicParam 1
Label :2	Return BasicParam 0

_getint: 
Label :5	Temp4 = call _getchar , 0
_ch = Temp4
Temp3 = _ch
BasicParam Temp3
Temp5 = call _isdecdigit , 1
If Not Equal to Zero Temp5 Goto Label :4	
Goto label:6
Label :6	Goto label:5
Goto label:5
Label :4	Temp7 = _ch SUB 48
_tmp = Temp7
Temp6 = _tmp
Label :8	Temp9 = call _getchar , 0
_ch = Temp9
Temp8 = _ch
BasicParam Temp8
Temp10 = call _isdecdigit , 1
If Not Equal to Zero Temp10 Goto Label :9	
Goto label:7
Label :9	Temp12 = _tmp MUL 10
Temp13 = Temp12 ADD _ch
Temp14 = Temp13 SUB 48
_tmp = Temp14
Temp11 = _tmp
Goto label:8
Label :7	Return BasicParam _tmp

_explode: 
If Not Equal to Zero flag Goto Label :11	
Goto label:10
Label :11	_i = 0
Temp15 = _i
Label :13	Temp16 = _i LT 1000
If Not Equal to Zero Temp16 Goto Label :14	
Goto label:12
Label :14	BasicParam Temp17
BasicParam _i
_printf_cnt = 2
Temp18 = call _printf , 2
Temp20 = _i ADD 1
_i = Temp20
Temp19 = _i
Goto label:13
Label :12	Label :10	Return BasicParam 1

_main: 
Temp21 = call _getint , 0
Temp22 = Temp21 ADD 1
_t1 = Temp22
Temp23 = call _getint , 0
Temp24 = Temp23 ADD 2
_t2 = Temp24
Temp25 = call _getint , 0
Temp26 = Temp25 ADD 3
_t3 = Temp26
_i = 100
Label :16	Temp27 = _i GT 0
If Not Equal to Zero Temp27 Goto Label :17	
Goto label:15
Label :17	Temp28 = _t1 GT 0
If Not Equal to Zero Temp28 Goto Label :19	
Goto label:21
Label :21	BasicParam _t1
Temp29 = call _explode , 1
If Not Equal to Zero Temp29 Goto Label :19	
Goto label:20
Label :19	Temp30 = _t2 GT 1
If Not Equal to Zero Temp30 Goto Label :26	
Goto label:25
Label :26	Temp31 = _t3 GT 2
If Not Equal to Zero Temp31 Goto Label :23	
Goto label:25
Label :25	BasicParam _t2
Temp32 = call _explode , 1
If Not Equal to Zero Temp32 Goto Label :23	
Goto label:24
Label :23	Temp33 = _t2 LE 0
If Not Equal to Zero Temp33 Goto Label :30	
Goto label:29
Label :30	BasicParam _t3
Temp34 = call _explode , 1
If Not Equal to Zero Temp34 Goto Label :28	
Goto label:29
Label :28	BasicParam _t3
Temp35 = call _explode , 1
Goto label:27
Label :29	If Not Equal to Zero _t2 Goto Label :38	
Goto label:37
Label :38	If Not Equal to Zero _t3 Goto Label :35	
Goto label:37
Label :37	BasicParam 1000
Temp36 = call _explode , 1
If Not Equal to Zero Temp36 Goto Label :35	
Goto label:36
Label :36	If Not Equal to Zero _t1 Goto Label :35	
Goto label:33
Label :35	If Not Equal to Zero 0 Goto Label :34	
Goto label:33
Label :34	BasicParam 1000
Temp37 = call _explode , 1
If Not Equal to Zero Temp37 Goto Label :32	
Goto label:33
Label :32	BasicParam 1000
Temp38 = call _explode , 1
Goto label:31
Label :33	BasicParam 0
Temp39 = call _explode , 1
If Not Equal to Zero Temp39 Goto Label :41	
Goto label:39
Label :41	Temp40 = _i EQ 1
If Not Equal to Zero Temp40 Goto Label :40	
Goto label:39
Label :40	BasicParam Temp41
BasicParam _t1
_printf_cnt = 2
Temp42 = call _printf , 2
Label :39	Label :31	Label :27	Goto label:22
Label :24	BasicParam _t2
Temp43 = call _explode , 1
Label :22	Goto label:18
Label :20	BasicParam _t1
Temp44 = call _explode , 1
Label :18	Temp46 = _i SUB 1
_i = Temp46
Temp45 = _i
Goto label:16
Label :15	Return BasicParam 0

